<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>“Advanced” Linux | Coinvasion Bioinformatics 2022</title>
<meta name="generator" content="Jekyll v3.9.2" />
<meta property="og:title" content="“Advanced” Linux" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Bioinformatics course for Coinvasion workshop 2022" />
<meta property="og:description" content="Bioinformatics course for Coinvasion workshop 2022" />
<link rel="canonical" href="http://localhost:4000/Linux/Advance" />
<meta property="og:url" content="http://localhost:4000/Linux/Advance" />
<meta property="og:site_name" content="Coinvasion Bioinformatics 2022" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="“Advanced” Linux" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"Bioinformatics course for Coinvasion workshop 2022","headline":"“Advanced” Linux","url":"http://localhost:4000/Linux/Advance"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="Coinvasion Bioinformatics 2022" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Coinvasion Bioinformatics 2022</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/">Home</a><a class="page-link" href="/Linux">Linux course</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post">

  <header class="post-header">
    <h1 class="post-title">&quot;Advanced&quot; Linux</h1>
  </header>

  <div class="post-content">
    <p>Notes:</p>

<p>Table related
cut: csv and tsv
more sort: -k 
awk</p>

<p>Crazy stuff:
Process substitution and command substitution</p>

<h2 id="loops-and-ifs">Loops and ifs</h2>
<h3 id="for">for</h3>
<p>We are using BLAST to learn about loops. And we will be using your favorite genes. First, we will build multiple databases with <code class="language-plaintext highlighter-rouge">for</code> loops and blast each databases.<br />
Let’s go to folder <code class="language-plaintext highlighter-rouge">LinuxAdvance/forloops</code> and copy three of your genome assemblies into the folder.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for f in *.fasta
do  makeblastdb -dbtype nucl -in $f
done
</code></pre></div></div>

<p>Let’s see what’s going on here. The first line uses a wild card and so it’s basically <code class="language-plaintext highlighter-rouge">for f in A.fasta B.fasta C.fasta</code>. This will create a variable call <code class="language-plaintext highlighter-rouge">f</code>. <code class="language-plaintext highlighter-rouge">A.fasta</code> will be assigned to <code class="language-plaintext highlighter-rouge">f</code> for the first iteration, <code class="language-plaintext highlighter-rouge">B.fasta</code> will be the next, and so on and so forth.<br />
Move on to the next line. <code class="language-plaintext highlighter-rouge">do</code> states the things you are going to do for an iteration. And <code class="language-plaintext highlighter-rouge">$f</code> is calling the variable <code class="language-plaintext highlighter-rouge">f</code>. So, for the first iteration, we are running <code class="language-plaintext highlighter-rouge">makeblastdb -dbtype nucl A.fasta</code>. So on and so forth.
Note that you can have multiple lines after do and the for loop will run through all of them. For example: (Don’t run)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for f in *.fasta
do  echo making database for $f
    makeblastdb -dbtype nucl -in $f
done
</code></pre></div></div>

<p>How about you try to run blast through all databases now? Copy the fasta with genes of interest into the folder and write a code to complete the task!</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  You can run something like this. (replace <code>blastn</code> with <code>tblastn</code> if you have a protein query.)
  <pre>for f in *.fasta  
do  blastn -db $f -query [&lt;yourquery&gt;] -outfmt 6 -out blastn-$f.txt  
done</pre>
</details>
<p><br /></p>

<p>Did you get an issue with your query fasta being treated as a database? There are at least three strategies to deal with it. The first one is to avoid <code class="language-plaintext highlighter-rouge">for</code> assigning it your variable <code class="language-plaintext highlighter-rouge">f</code>. For example, you can rename it, put it into a different folder or use some clever scheme in your wildcard to exclude the query.<br />
Or, you can use <code class="language-plaintext highlighter-rouge">if</code> (I’m still proposing using the former, since it will be cleaner).</p>

<details>
  <summary><b><u>Click to learn more!</u></b></summary>
  <code>for</code> loops are also useful if you want to generate a series of number. Try the followings:  
  <pre>for n in {1..20}  
do  echo $n
done</pre>
</details>
<p><br /></p>

<h3 id="if-else">if else</h3>
<p>I’m not using <code class="language-plaintext highlighter-rouge">if</code> <code class="language-plaintext highlighter-rouge">else</code> a lot, but I think it’s worth mentioning.
Using our former task as example, we want to avoid your query. Therefore we need a <code class="language-plaintext highlighter-rouge">if</code> clause saying “if <code class="language-plaintext highlighter-rouge">f</code> is query, skip it”, or “if <code class="language-plaintext highlighter-rouge">f</code> is not query, do some stuff”.<br />
Let’s start with the second strategy with an example.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x=somestring
if [ $x != somestring ]
then echo $x is not somestring
fi
</code></pre></div></div>

<p>In this example, we first assign “somestring” to <code class="language-plaintext highlighter-rouge">x</code>. The second line ask if <code class="language-plaintext highlighter-rouge">x</code> is not somestring (<code class="language-plaintext highlighter-rouge">!=</code> is not equal). Now, the <code class="language-plaintext highlighter-rouge">[]</code> pass a true to if, so the if clause continue to run things from <code class="language-plaintext highlighter-rouge">then</code> to <code class="language-plaintext highlighter-rouge">fi</code>. You can try assign something else to <code class="language-plaintext highlighter-rouge">x</code> and see what happens.
Note: You need the space between <code class="language-plaintext highlighter-rouge">[</code> and <code class="language-plaintext highlighter-rouge">$x</code> and between <code class="language-plaintext highlighter-rouge">somestring</code> and <code class="language-plaintext highlighter-rouge">]</code>.</p>

<p>Now, let’s try the first strategy.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x=somestring
if [ $x == somestring ]
then echo
else echo $x is not somestring
fi
</code></pre></div></div>

<p>Here, there’s an else clause to say “if the things in <code class="language-plaintext highlighter-rouge">[]</code> is false, in this case: <code class="language-plaintext highlighter-rouge">x</code> is somestring (<code class="language-plaintext highlighter-rouge">==</code> is equal), run some <code class="language-plaintext highlighter-rouge">echo $x is not somestring</code>.<br />
You can see the solution is not elegant. You still need to run something when <code class="language-plaintext highlighter-rouge">[]</code> is true (I just gave it an <code class="language-plaintext highlighter-rouge">echo</code>). So, avoid this solution, but <code class="language-plaintext highlighter-rouge">else</code> is useful when you want to run something dichotomous.</p>

<details>
  <summary><b><u>Click to learn more!</u></b></summary>
  There's also <code>elif</code> so you can process multiple if else together. Go learn it yourself!
</details>
<p><br /></p>

<p>Try to write a code with <code class="language-plaintext highlighter-rouge">for</code> and <code class="language-plaintext highlighter-rouge">if</code> to avoid using query fasta!</p>
<details>
  <summary><b><u>Hint</u></b></summary>
  <pre>for f in *.fasta  
do  if [ $f != [&lt;yourquery&gt;] ]  
    then blastn -db $f -query [&lt;yourquery&gt;] -outfmt 6 -out tblastn-$f.txt  
    fi  
done</pre>
</details>
<p><br /></p>

<h3 id="while-loops">while loops</h3>
<p>The third strategy is to use <code class="language-plaintext highlighter-rouge">while</code> loops.<br />
The most common usage of <code class="language-plaintext highlighter-rouge">while</code> loop is to read a file and it will go through each line and assign them to variables.
Let’s go to <code class="language-plaintext highlighter-rouge">LinuxAdvance/forloops</code>, take a look at <code class="language-plaintext highlighter-rouge">testfile</code> and run this:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>while read l
do  echo $l is a letter
done &lt; testfile
</code></pre></div></div>

<p>Could you figure out what’s going on?</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  The first line saying while you are reading a line from a file, you assign it to <code>l</code> and do the things between <code>do</code> and <code>done</code>.  
  The last line uses <code>&lt;</code> to indicate which file you are reading.
</details>
<p><br /></p>

<p>Now, your turn, copy three of your genome assemblies and the fasta with genes of interest into the folder. Then, write a code to complete the task.
Note: use <code class="language-plaintext highlighter-rouge">nano</code> to create the files that you are reading with <code class="language-plaintext highlighter-rouge">while</code>.</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  It should be something like this:
  <pre>while read l
do makeblastdb -dbtype nucl $l
  blastn -db $l -query [&lt;yourquery&gt;] -outfmt 6 -out tblastn-$l.txt  
done &lt; [&lt;yourfastalist&gt;] </pre>
Also see that I combined makeblastdb and blast so I don't need two loops! 
</details>
<p><br /></p>

<details>
  <summary><b><u>Click to learn more!</u></b></summary>
  There's also <code>elif</code> so you can process multiple if else together. Go learn it yourself!
</details>
<p><br /></p>

<h2 id="advance-sed-and-grep">Advance sed and grep</h2>
<h3 id="sed-with-variables">sed with variables</h3>
<p>To make a combined database with multiple genomes, we technically only need to concatenate all the genome files and then use <code class="language-plaintext highlighter-rouge">makeblastdb</code>. However, it would be hard to know which contig or scaffold is from which sample. So, we are going back to <code class="language-plaintext highlighter-rouge">sed</code> to label these contigs/scaffolds. And although we do not necessarily need complicated coding to accomplish it, it’s a good point to introduce you regular expressions. Just bear with me.</p>

<p>If we look at our assemblies, each sequence have a sequence name start with <code class="language-plaintext highlighter-rouge">&gt;</code>. The easiest way to add sample information is to replace <code class="language-plaintext highlighter-rouge">&gt;</code> with <code class="language-plaintext highlighter-rouge">&gt;[&lt;sample name&gt;]</code>.<br />
Note that because you are using <code class="language-plaintext highlighter-rouge">''</code> in <code class="language-plaintext highlighter-rouge">sed</code> and that will suppress the meaning of <code class="language-plaintext highlighter-rouge">$</code> for calling variables, so you’ll need to use <code class="language-plaintext highlighter-rouge">''</code> to address it. Basically, if you want to replace variable <code class="language-plaintext highlighter-rouge">X</code> with variable <code class="language-plaintext highlighter-rouge">Y</code>, you should code</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/'$X'/'$Y'/g' [&lt;somefile&gt;]
</code></pre></div></div>

<p>Now, you should have all the things you need to figure it out!<br />
Let’s go to <code class="language-plaintext highlighter-rouge">LinuxAdvance/sedblast</code>, copy a few assembled genomes there and try it out! (Don’t stop at making database but also do the blast part.)</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  I'm skipping the <code>makeblastdb</code> and <code>blast</code>
  <pre>for f in *.fasta  
do  sed 's/&gt;/&gt;'$f'/g' $f &gt; temp  
    mv temp $f
done  
cat *.fasta &gt; compiled.fasta  </pre>
  You will get something that's sort of usable with this script.<br />
  But it's ugly. you will have <code>.fasta</code> inside your sequences. What can you do? Well, just pipe it to another <code>sed</code> to remove it (and maybe replace it with something sensible!)
</details>
<details>
  <summary><b><u>Click to learn more!</u></b></summary>
  You might find making a temporary file very inelegant. You can check out <code>-i</code>. I just never remember to use it :).
</details>
<p><br /></p>

<p>Take a look at the blast results what did you find?</p>

<h3 id="sed-with-regular-expression-regex">sed with regular expression (Regex)</h3>

<h4 id="beginning-and-end-of-a-line">Beginning and end of a line.</h4>

<p>Normally you won’t have this problem, but what if there’s an additional <code class="language-plaintext highlighter-rouge">&gt;</code> in the middle of the sequence name? That will also add in the sample name that you don’t need right?</p>

<p>So, we can use <code class="language-plaintext highlighter-rouge">^</code> to specify that <code class="language-plaintext highlighter-rouge">&gt;</code> needs to be at the beginning. Like:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for f in *.fasta  
do  sed 's/^&gt;/&gt;'$f'/g' $f &gt; temp  
    mv temp $f
done 
</code></pre></div></div>

<p>Sometimes you also want to add something to the end of a sequence. You can then use <code class="language-plaintext highlighter-rouge">$</code> to specify.
For example: (do not run)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/A$/B/g' 
</code></pre></div></div>

<p>This will replace every <code class="language-plaintext highlighter-rouge">A</code> at end of a line with <code class="language-plaintext highlighter-rouge">B</code>.</p>

<p>I hope that you are interested in Regex now. And we will learn some other useful Regex!</p>

<h4 id="wildcards">Wildcards</h4>

<p>Let’s head to <code class="language-plaintext highlighter-rouge">LinuxAdvance/sedregex</code>, take a look at the file <code class="language-plaintext highlighter-rouge">ECMspecies</code> and learn about more Regex.<br />
First we will learn <code class="language-plaintext highlighter-rouge">.</code>. <code class="language-plaintext highlighter-rouge">.</code> means any single character. Let’s see what happen if we run these two scripts:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/./XXX/' ECMspecies
sed 's/../XXX/' ECMspecies
</code></pre></div></div>

<p>Notice that I didn’t add <code class="language-plaintext highlighter-rouge">g</code> at the end of the <code class="language-plaintext highlighter-rouge">sed</code> expression, so it will only replace the first match.</p>

<p><code class="language-plaintext highlighter-rouge">*</code> is often used with <code class="language-plaintext highlighter-rouge">.</code>. What it does is matching the preceding element zero or more times (It’s different from <code class="language-plaintext highlighter-rouge">bash</code>’s wildcard.)</p>

<p>So what does this do?</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/ .*/XXX/g' ECMspecies
</code></pre></div></div>

<p>Let’s try to add <code class="language-plaintext highlighter-rouge">feminine</code> after genus names that end with <code class="language-plaintext highlighter-rouge">a</code>, add <code class="language-plaintext highlighter-rouge">masculine</code> after genus names that end with <code class="language-plaintext highlighter-rouge">us</code>. Also, remove the numbers and species. (Ignore <code class="language-plaintext highlighter-rouge">Tuber</code>)</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  <pre>sed 's/a .*/a feminine/g' ECMspecies | sed 's/us .*/us masculine/g'  </pre>
</details>
<p><br /></p>

<h4 id="bracket-expressions">Bracket expressions</h4>

<p>Although <code class="language-plaintext highlighter-rouge">.</code> is useful, we can’t do more nuanced edits. For example, it can not do something like only editting <code class="language-plaintext highlighter-rouge">a</code> and <code class="language-plaintext highlighter-rouge">c</code> in the file. But, we can use bracket expressions <code class="language-plaintext highlighter-rouge">[]</code>.</p>

<p>How bracket expression works is that you place everything that you want to match inside the bracket. Such as,</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/[ac]/XXX/' ECMspecies
</code></pre></div></div>

<p>Notice that, again, I did not use <code class="language-plaintext highlighter-rouge">g</code>. Take a look at the results. What happened?</p>

<p>We can also use <code class="language-plaintext highlighter-rouge">-</code> to replace a series of characters or numbers. For example,</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/[a-c]/XXX/' ECMspecies
sed 's/[0-3]/XXX/' ECMspecies
</code></pre></div></div>

<p>What just happened then? Look at the <code class="language-plaintext highlighter-rouge">Boletus</code> line. Found anything?</p>

<p>You can also specify the amount of times of these brackets need to appear with <code class="language-plaintext highlighter-rouge">\{\}</code>. Let’s look at how it works without brackets first.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/l{2}/XXX/' ECMspecies
</code></pre></div></div>

<p>What does this do?</p>

<p>You can replace the <code class="language-plaintext highlighter-rouge">l</code> with brackets. Like,</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/[ac]\{2\}/XXX/' ECMspecies
</code></pre></div></div>

<p>You can also specify a range of number of characters. Like:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/[lsu]\{1,\}/XXX/' ECMspecies
sed 's/[lsu]\{1,3\}/XXX/' ECMspecies
</code></pre></div></div>

<p>What does each of these do?</p>

<p>Now, you should know how to replace the numbers only. Let’s try to replace them with <code class="language-plaintext highlighter-rouge">N</code>!</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  <pre>sed 's/[0-9]\{1,\}/N/' ECMspecies  </pre>
</details>

<p><br /></p>

<p>Lastly, you should also know that you can add <code class="language-plaintext highlighter-rouge">^</code> at the beginning of the brackets to indicate “not”.
For example,</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/[^0-9]/X/g' ECMspecies
</code></pre></div></div>
<p>What does it do?</p>

<p><br /></p>

<h4 id="marked-subexpression">Marked subexpression</h4>

<p>Now, let’s say we want move the numbers to the beginning of the line, we can use marked subexpression <code class="language-plaintext highlighter-rouge">\(\)</code>.</p>

<p>Run the followings:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>echo ABCDEFGH | sed 's/^\(.*\)\(F\)/\2\1/'
</code></pre></div></div>

<p>What does it do? Let’s ignore the <code class="language-plaintext highlighter-rouge">\(\)</code> first.<br />
<code class="language-plaintext highlighter-rouge">^</code> indicates the match needs to start from the beginning. After <code class="language-plaintext highlighter-rouge">.*</code> we find <code class="language-plaintext highlighter-rouge">F</code> and we replace them with <code class="language-plaintext highlighter-rouge">\2</code> and <code class="language-plaintext highlighter-rouge">\1</code>.<br />
<code class="language-plaintext highlighter-rouge">\1</code> and <code class="language-plaintext highlighter-rouge">\2</code> are defined by <code class="language-plaintext highlighter-rouge">\(\)</code>. <code class="language-plaintext highlighter-rouge">\1</code> corresponds to the first and <code class="language-plaintext highlighter-rouge">\2</code> correspond to the second.</p>

<p>If it makes sense, try to move the numbers to the beginning of the line like this:</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  <pre>sed 's/\(.*\) \([0-9]\{1,\}\)/\2 \1/' ECMspecies  </pre>
</details>

<p><br /></p>

<h4 id="grep">grep</h4>
<p><code class="language-plaintext highlighter-rouge">grep</code> also supports the usage of Regex. Are you able to find all the genera with species number equal or higher than 10 but lower than 100? (10~99)</p>

<details>
  <summary><b><u>Hint</u></b></summary>
  <pre>grep ' [0-9]\{2\} ' ECMspecies  </pre>
</details>

<p><br /></p>

<details>
  <summary><b><u>Click to learn more!</u></b></summary>
  The Regex we used was POSIX basic regex. There's also a thing called POSIX extended basic regex. You can use <code>-E</code> to enable it. You don't need <code>\</code> for <code>{}</code> and <code>()</code> if you are using the extended version. And there are a few other functions that you can use. Learn more [here](https://en.wikipedia.org/wiki/Regular_expression)!
</details>

<p><br /></p>

<p><img src="/img/band.png" alt="band" /></p>

<p><br /></p>

<h2 id="dealing-with-tables">dealing with tables</h2>

<h3 id="cut">cut</h3>

<h3 id="sort">sort</h3>

<p>cut: csv and tsv
more sort: -k 
awk</p>

<p>cut</p>

<p>sort</p>

<p>awk!</p>


  </div>

</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">Coinvasion Bioinformatics 2022</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Coinvasion Bioinformatics 2022</li><li><a class="u-email" href="mailto:yen-wen.wang@yale.edu">yen-wen.wang@yale.edu</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>Bioinformatics course for Coinvasion workshop 2022</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
